# <span style="color: Orange;"> Credit Card Transaction Analysis - PART 1: Credit  card company POV </span>

### *<span style= "color:gray;"> Thought process, Insights and Recommendations</span>*

### *Context* :  
 __Bank "ABC" had inaugurated offering their credit card facility on 1st Oct, 2013.__  
__They started collecting data on the credit card transaction details since then.__  
    
__Now, after nearly two years, they have a stakeholder meeting on June 1st 2015 to assess the situation regarding the growth in the credit card section.__  

__The data till now (latest data being 26th May 2015) needs to be analysed and <span style="color:green;">insights and recommendations needs to be presented to the stakeholders during their meeting so that they can take data-driven decision on how to proceed.__</span>

### Source of data:  
*Credit_card_transactions_India.csv*

### *<span style= "color:gray;"> STEP 1: Getting to know the data</span>*

<span style="font-size:14px;"> __Let's start with getting the look and feel of the data that we need to work with.__ 
    
__We would be looking at the number of rows and columns of the data along with what each column denotes and the data types.__ 
    
__We would also be noting various obvious issues/ anomalies with the data that needs to be fixed as a part of cleaning the data although data cleaning is not the mail goal of this section.__ </span>

In [72]:
import numpy as np
import pandas as pd

In [73]:
original_df = pd.read_csv(r"D:\Data Analysis\Datasets\Credit_card_transactions_India.csv")

In [74]:
working_df = original_df.copy(deep=True)
working_df

index                   City       Date  Card Type Exp Type Gender  \
0          0           Delhi, India  29-Oct-14       Gold    Bills      F   
1          1  Greater Mumbai, India  22-Aug-14   Platinum    Bills      F   
2          2       Bengaluru, India  27-Aug-14     Silver    Bills      F   
3          3  Greater Mumbai, India  12-Apr-14  Signature    Bills      F   
4          4       Bengaluru, India   5-May-15       Gold    Bills      F   
...      ...                    ...        ...        ...      ...    ...   
26047  26047         Kolkata, India  22-Jun-14     Silver   Travel      F   
26048  26048            Pune, India   3-Aug-14  Signature   Travel      M   
26049  26049       Hyderabad, India  16-Jan-15     Silver   Travel      M   
26050  26050          Kanpur, India  14-Sep-14     Silver   Travel      M   
26051  26051       Hyderabad, India  19-Oct-13  Signature    Bills      M   

       Amount  
0       82475  
1       32555  
2      101738  
3      123424  
4      171574  
...       ...  
26047  128191  
26048  246316  
26049  265019  
26050   88174  
26051  184410  

[26052 rows x 7 columns]


- __There are 6 columns (excluding index) and 26k rows__ 
- __First things first, the index column provided in the file needs to be dropped.__ 
- __The city column has the format: "city name", India.  
    We can drop the comma and India, as it doesn't add any value to the analysis and just retain the actual city name.__ 
- __The dates don't seem to be in chronological order, so the dataframe has to be sorted with dates in ascending order.__
    



    
### <span style= "color:gray;"> Column_name: Description (expected data type ): </span>

*City: The city in which the transaction took place. (String)* 

*Date: The date of the transaction. (Date)*  

*Card Type: The type of credit card used for the transaction. (String)*  

*Exp Type: The type of expense associated with the transaction. (String)*  

*Gender: The gender of the cardholder. (String)*  

*Amount: The amount of the transaction. (Number)*

__Before we do any further analysis, let's correct the obvious issues__

###  <span style="color:red;">Let's start by dropping the in-built index column </span>

In [75]:
# Dropping in-built Index column
working_df = working_df.drop('index', axis=1)
working_df

City       Date  Card Type Exp Type Gender  Amount
0               Delhi, India  29-Oct-14       Gold    Bills      F   82475
1      Greater Mumbai, India  22-Aug-14   Platinum    Bills      F   32555
2           Bengaluru, India  27-Aug-14     Silver    Bills      F  101738
3      Greater Mumbai, India  12-Apr-14  Signature    Bills      F  123424
4           Bengaluru, India   5-May-15       Gold    Bills      F  171574
...                      ...        ...        ...      ...    ...     ...
26047         Kolkata, India  22-Jun-14     Silver   Travel      F  128191
26048            Pune, India   3-Aug-14  Signature   Travel      M  246316
26049       Hyderabad, India  16-Jan-15     Silver   Travel      M  265019
26050          Kanpur, India  14-Sep-14     Silver   Travel      M   88174
26051       Hyderabad, India  19-Oct-13  Signature    Bills      M  184410

[26052 rows x 6 columns]

In [76]:
#Sorting chronologically
#working_df['Date'] = pd.to_datetime(working_df['Date'])
working_df = working_df.sort_values(by='Date', ascending=True)
working_df

City      Date  Card Type       Exp Type Gender  Amount
24582       Chennai, India  1-Apr-14   Platinum  Entertainment      F   93038
16742  Wadgaon Road, India  1-Apr-14       Gold           Food      M  274661
4305      Bengaluru, India  1-Apr-14   Platinum          Bills      F  174903
20368        Jaipur, India  1-Apr-14   Platinum  Entertainment      M  118570
15309         Nokha, India  1-Apr-14  Signature          Bills      M  163677
...                    ...       ...        ...            ...    ...     ...
12607     Bengaluru, India  9-Sep-14       Gold           Food      M  259100
3136          Delhi, India  9-Sep-14   Platinum           Food      F  217217
17742         Udupi, India  9-Sep-14       Gold  Entertainment      M  105855
7435          Delhi, India  9-Sep-14  Signature           Food      M   46182
15194  Rajagangapur, India  9-Sep-14  Signature           Food      M  165047

[26052 rows x 6 columns]

###  <span style="color:maroon;">The above result shows the importance of making sure the date column has the correct data type </span>  

Let's check the data types of these columns

In [77]:
working_df.dtypes

City         object
Date         object
Card Type    object
Exp Type     object
Gender       object
Amount        int64
dtype: object

 #### <span style="color:red;">  data type of "Date" column is object. </span>  
 
 ### <span style="color:red;"> Let's change it to datetime and see how the new sorting is</span>  

In [78]:
# changing date column data type

working_df['Date'] = pd.to_datetime(working_df['Date'])
working_df.dtypes

City                 object
Date         datetime64[ns]
Card Type            object
Exp Type             object
Gender               object
Amount                int64
dtype: object

In [79]:
#Sorting chronologically after changing date column data type

working_df = working_df.sort_values(by='Date', ascending=True)
working_df

City       Date  Card Type       Exp Type Gender  \
72              Delhi, India 2013-10-04  Signature          Bills      F   
7201   Greater Mumbai, India 2013-10-04       Gold  Entertainment      M   
6931        Bengaluru, India 2013-10-04       Gold           Fuel      F   
8645        Bengaluru, India 2013-10-04     Silver          Bills      M   
23691         Chennai, India 2013-10-04       Gold        Grocery      M   
...                      ...        ...        ...            ...    ...   
7891        Ahmedabad, India 2015-05-26   Platinum        Grocery      F   
24604          Jaipur, India 2015-05-26     Silver           Fuel      F   
9980            Delhi, India 2015-05-26   Platinum  Entertainment      F   
22648         Lucknow, India 2015-05-26       Gold        Grocery      M   
14658          Washim, India 2015-05-26  Signature           Fuel      F   

       Amount  
72     550782  
7201   201942  
6931   236037  
8645   159769  
23691   27511  
...       ...  
7891   269998  
24604  292561  
9980   200932  
22648  249011  
14658  223068  

[26052 rows x 6 columns]

__Glancing at the datframe output:__

__Oldest entry (first entry) is on 4th Oct 2013 (2013-10-04) and__ 

__Latest entry (last entry) is on 26th May 2015 (2015-05-26)__

### <span style="color:red;"> Let's check the oldest and latest entry to confirm that the sorting is correct. </span>

In [80]:
# checking the oldest entry and the latest entry
#dt.date extracts just the date part of the datetime value

oldest_entry = working_df["Date"].dt.date.min() 
latest_entry = working_df["Date"].dt.date.max()
print("oldest entry: ", oldest_entry)
print("latest entry: ", latest_entry)

oldest entry:  2013-10-04
latest entry:  2015-05-26


__sorting seems to have worked properly as the oldest and latest values cross-checks with the dates visible in the dataframe output__

 ### <span style="color:red;">  but now the index is all jumbled, so let's reset it. </span> 
Note that on resetting, the current index would be added as a new column, hence we need to set drop=True for dropping the older, jumbled index

In [81]:
# Resetting index
working_df = working_df.reset_index(drop=True)
working_df

City       Date  Card Type       Exp Type Gender  \
0               Delhi, India 2013-10-04  Signature          Bills      F   
1      Greater Mumbai, India 2013-10-04       Gold  Entertainment      M   
2           Bengaluru, India 2013-10-04       Gold           Fuel      F   
3           Bengaluru, India 2013-10-04     Silver          Bills      M   
4             Chennai, India 2013-10-04       Gold        Grocery      M   
...                      ...        ...        ...            ...    ...   
26047       Ahmedabad, India 2015-05-26   Platinum        Grocery      F   
26048          Jaipur, India 2015-05-26     Silver           Fuel      F   
26049           Delhi, India 2015-05-26   Platinum  Entertainment      F   
26050         Lucknow, India 2015-05-26       Gold        Grocery      M   
26051          Washim, India 2015-05-26  Signature           Fuel      F   

       Amount  
0      550782  
1      201942  
2      236037  
3      159769  
4       27511  
...       ...  
26047  269998  
26048  292561  
26049  200932  
26050  249011  
26051  223068  

[26052 rows x 6 columns]

 ### <span style="color:red;"> Now let's split the City column. </span>

In [82]:
working_df[['City', 'Country']] = working_df['City'].str.split(',', expand=True)
working_df

City       Date  Card Type       Exp Type Gender  Amount  \
0               Delhi 2013-10-04  Signature          Bills      F  550782   
1      Greater Mumbai 2013-10-04       Gold  Entertainment      M  201942   
2           Bengaluru 2013-10-04       Gold           Fuel      F  236037   
3           Bengaluru 2013-10-04     Silver          Bills      M  159769   
4             Chennai 2013-10-04       Gold        Grocery      M   27511   
...               ...        ...        ...            ...    ...     ...   
26047       Ahmedabad 2015-05-26   Platinum        Grocery      F  269998   
26048          Jaipur 2015-05-26     Silver           Fuel      F  292561   
26049           Delhi 2015-05-26   Platinum  Entertainment      F  200932   
26050         Lucknow 2015-05-26       Gold        Grocery      M  249011   
26051          Washim 2015-05-26  Signature           Fuel      F  223068   

      Country  
0       India  
1       India  
2       India  
3       India  
4       India  
...       ...  
26047   India  
26048   India  
26049   India  
26050   India  
26051   India  

[26052 rows x 7 columns]

 ### <span style="color:red;"> We can remove the country column as it adds no value for analysis. </span>
 
 #### <span style="color:red;">But wait, are we sure it only contains India? </span> 
 
 Let's check the unique values in that column to confirm before dropping

In [83]:
# Checking the unique entries in "Country" column

working_df['Country'].unique()

array([' India'], dtype=object)

__Looks good as it only seems to contain India.__

### <span style="color:red;"> Let's drop "Country" column. </span>

In [84]:
# dropping country column

working_df = working_df.drop('Country', axis=1)
working_df

City       Date  Card Type       Exp Type Gender  Amount
0               Delhi 2013-10-04  Signature          Bills      F  550782
1      Greater Mumbai 2013-10-04       Gold  Entertainment      M  201942
2           Bengaluru 2013-10-04       Gold           Fuel      F  236037
3           Bengaluru 2013-10-04     Silver          Bills      M  159769
4             Chennai 2013-10-04       Gold        Grocery      M   27511
...               ...        ...        ...            ...    ...     ...
26047       Ahmedabad 2015-05-26   Platinum        Grocery      F  269998
26048          Jaipur 2015-05-26     Silver           Fuel      F  292561
26049           Delhi 2015-05-26   Platinum  Entertainment      F  200932
26050         Lucknow 2015-05-26       Gold        Grocery      M  249011
26051          Washim 2015-05-26  Signature           Fuel      F  223068

[26052 rows x 6 columns]

### Now we are ready for some more EDA

### <span style="color:red;"> Let's check how many unique cities are there. </span>

In [85]:
# Number of unique cities and their names

print('Number of unique cities: ', len(working_df['City'].unique()))

Number of unique cities:  986


### <span style="color:purple;"> printing out all 986 names wouldn't help much, hence let's skip that step. </span>

### <span style="color:red;"> Let's check how many unique card types are there. </span>

In [86]:
# Number of unique card types and their names

print('Number of unique card types: ', len(working_df['Card Type'].unique()))

Number of unique card types:  4


In [87]:
print('Name of unique card types: \n', working_df['Card Type'].unique())

Name of unique card types: 
 ['Signature' 'Gold' 'Silver' 'Platinum']


### <span style="color:purple;"> The possible hierarchy would be: </span>  

- __Signature__  

- __Platinum__  

- __Gold__  

- __Silver__

*Assumed from the fact that Visa is the most notable providers of Signature card type which is above their Platinum*

### <span style="color:red;"> Let's check how many unique expense types (Exp Type) are there. </span>

In [88]:
# Number of unique exp types and their names

print('Number of unique exp types: ', len(working_df['Exp Type'].unique()))

Number of unique exp types:  6


In [89]:
print('Name of unique exp types: \n', working_df['Exp Type'].unique())

Name of unique exp types: 
 ['Bills' 'Entertainment' 'Fuel' 'Grocery' 'Food' 'Travel']


### <span style="color:red;"> Let's check how many unique genders are there. </span>

#### <span style="color:red;"> We cannot assume this and possibly overlook the cases of Non-binary.</span>

In [90]:
# Number of unique genders and their names

print('Number of unique genders: ', len(working_df['Gender'].unique()))

Number of unique genders:  2


In [91]:
print('Name of unique genders: \n', working_df['Gender'].unique())

Name of unique genders: 
 ['F' 'M']


### <span style="color:red;"> Let's check the summary statistics for the 'Amount' column </span> 

In [92]:
# summary statistics for the 'Amount' column
# given that amount is the only numeric column we can do it directly on the dataframe

working_df.describe()

Amount
count   26052.000000
mean   156411.537425
std    103063.254287
min      1005.000000
25%     77120.250000
50%    153106.500000
75%    228050.000000
max    998077.000000

### <span style="color:red;"> Let's calculate the interquartile range </span> 

In [93]:
#calculating interquartile range 

Q1 = working_df['Amount'].quantile(q=0.25)
Q3 = working_df['Amount'].quantile(q=0.75)
IQR = Q3 - Q1
print("Lower quartile: ", Q1)
print("Upper quartile: ", Q3)
print("Interquartile range (IQR): ",IQR)

Lower quartile:  77120.25
Upper quartile:  228050.0
Interquartile range (IQR):  150929.75


### <span style="color:red;"> Let's calculate the lower bound and upper bound and define the outliers  </span> 

In [94]:
Lower_bound = Q1 -(1.5*IQR)
Upper_bound = Q3 + (1.5*IQR)
print("Lower bound: ", Lower_bound)
print("Upper bound: ", Upper_bound)


Lower bound:  -149274.375
Upper bound:  454444.625


#### <span style="color:purple;"> Since lower bound is negative and unrealistic, we would only be considering outliers for upper bound. </span>

### <span style="color:purple;"> An oulier in this case is any amount greater than upper bound (~ 454k)  </span>

# <span style="color:Teal;"> INFERENCE #1 </span> 

There is a lot to unpack here.  

> The count matches the row count indicating there are no missing values, which is good news.  

> The mean (156k) is slightly greater than median (153k) indicating that the distribution is right skewed, possiblilty of high transaction amounts and even outliers [note that 50% value is the median]  

> standard deviation is 103k which is quite high (almost two-thirds of the mean which is 156k) and this means the data points are spread out widely. That means there is a lot of variability in transaction amounts.

> IQR (Interquartile range) is from 77k to 228k (25% to 75%)

> The max value (998k) is more than 2 times the upper bound value (454k)

In [96]:
outliers = working_df[working_df['Amount'] >=454000]
outliers

City       Date  Card Type Exp Type Gender  Amount
0               Delhi 2013-10-04  Signature    Bills      F  550782
5      Greater Mumbai 2013-10-04       Gold    Bills      F  990700
6               Delhi 2013-10-04  Signature    Bills      F  735037
56          Ahmedabad 2013-10-05   Platinum    Bills      F  612572
126         Bengaluru 2013-10-06  Signature    Bills      F  677654
...               ...        ...        ...      ...    ...     ...
25633       Bengaluru 2015-05-17     Silver    Bills      F  792440
25714       Bengaluru 2015-05-19  Signature    Bills      F  977951
25782  Greater Mumbai 2015-05-20     Silver    Bills      F  567623
25837       Bengaluru 2015-05-22   Platinum    Bills      F  500852
25897  Greater Mumbai 2015-05-23  Signature    Bills      F  489043

[229 rows x 6 columns]

# <span style="color:Magenta;"> INSIGHTS #1 </span> 

- Risk assessment 

> High variability implies unpredictability. If transactions vary a lot, it's harder to predict future spending. 

> For a bank, this could mean higher risk because they can't reliably estimate how much customers might spend or repay.

> Also, when calculating the probability of default, the variability in spending could correlate with income volatility. Customers with erratic spending might have unstable income, increasing default risk.

> If some customers have very high transactions (outliers), they might be profitable but also pose higher risk because they might be more likely to default if those large amounts aren't managed properly. 

> fraud risk increases because unusually large transactions could be fraudulent, and the bank needs to detect those.   

> For risk assessment models, high variability might affect the accuracy. Models that assume normal distribution might not work well here.

- Credit Limits

> The bank needs to set limits that balance customer spending needs with risk. 

> Setting a uniform limit based on mean or median could be problematic.If there's high variability, using the mean might not be sufficient because some customers might need higher limits, but the bank has to balance that with the risk of non-payment. 

> The bank might need to adjust credit limits based on the variability. 